src/__tests__/useAsync-test.js

```diff
@@ -0,0 +1,156 @@
+/* eslint react/prop-types: 0 */
+import '@testing-library/jest-dom/extend-expect';
+import * as React from 'react';
+import '@babel/polyfill';
+import { render, waitForDomChange, fireEvent, act } from '@testing-library/react';
+import useAsync from '../useAsync';
+
+function DisplayState({ api }) {
+  const result = {
+    value: api.value(),
+    success: api.success(),
+    error: api.error(),
+    initial: api.initial(),
+    loading: api.loading(),
+
+    hasValue: api.hasValue(),
+    isSuccess: api.isSuccess(),
+    isLoading: api.isLoading(),
+    isError: api.isError(),
+    isInitial: api.isInitial(),
+
+    matches: api.match({
+      default: () => ({ type: 'default' }),
+      loading: () => ({ type: 'loading' }),
+      success: value => ({ type: 'success', value }),
+      error: msg => ({ type: 'error', value: msg.toString() }),
+    }),
+  };
+
+  return (
+    <div data-testid="state">
+      {JSON.stringify(result)}
+    </div>
+  );
+}
+
+function getStates(getByTestId) {
+  return JSON.parse(getByTestId('state').textContent);
+}
+
+it(`matches only loading`, () => {
+  const neverResolves = () => new Promise(() => {});
+
+  function Test() {
+    const api = useAsync(() => neverResolves(), []);
+    return <DisplayState api={api} />;
+  }
+
+  const { getByTestId } = render(<Test />);
+  expect(getStates(getByTestId)).toEqual({
+    value: undefined,
+    success: null,
+    error: null,
+    initial: false,
+    loading: true,
+
+    hasValue: false,
+    isSuccess: false,
+    isLoading: true,
+    isError: false,
+    isInitial: false,
+
+    matches: { type: 'loading' },
+  });
+});
+
+it(`resolves to success`, async () => {
+  const resolves = () => new Promise(resolve => resolve(1));
+
+  function Test() {
+    const api = useAsync(() => resolves(), []);
+    return <DisplayState api={api} />;
+  }
+
+  const { getByTestId } = render(<Test />);
+  await waitForDomChange();
+  expect(getStates(getByTestId)).toEqual({
+    value: 1,
+    success: 1,
+    error: null,
+    initial: false,
+    loading: false,
+
+    hasValue: true,
+    isSuccess: true,
+    isLoading: false,
+    isError: false,
+    isInitial: false,
+
+    matches: { type: 'success', value: 1 },
+  });
+});
+
+it(`rejects to error`, async () => {
+  const rejects = () => new Promise((_, reject) => reject(new Error('rejected')));
```

Comment:
Reject to something you can test against, e.g. 

```js
const error = { message: 'Expected rejection' };
const rejects = () => new Promise((_, reject) => reject(error));
```

And use `error` in the toEqual.